### ISSUE TITLE
BinaryDataUtilities.AsArray doesn't take into account count of ArraySegment

### ISSUE THREAD

**Issue by arekpalinski** [cloudevents/sdk-csharp#209]
Hi,

I had some problems with own implementation of `CloudEventFormatter` which caused that some of my JSON messages got corrupted in the kafka topic. They contained more data than expected. I narrowed it down to the following piece of code:

https://github.com/cloudevents/sdk-csharp/blob/8f75a9b3cb8cd2c52378d4e71ea64fb9a602eae2/src/CloudNative.CloudEvents/Core/BinaryDataUtilities.cs#L109-L110

It looks the issue is that _entire_ `segment.Array` might be returned even though it can have bigger size than the size of passed memory. Please check the following test:

```
byte[] array = new byte[100];

array[0] = 0;
array[1] = 1;
array[2] = 2;


var arraySegment = new ArraySegment<byte>(array, 0, 3);

var bytes = BinaryDataUtilities.AsArray(arraySegment);

Assert.Equal(array.Length, bytes.Length); // FAILS: bytes.Length is 100
```

**Comment by jskeet** [comment]
Ah - it looks like it should be `&&` instead of `||` - although I haven't looked at the full code, just the snippet in this issue. Thanks for the report. I'll try to get a new version out by the end of the week.

**Comment by jskeet** [comment]
Thank goodness for tests - that wasn't all that was required :)

I've created #210 for this, but I'd really love to get more validation that it's actually working. Are you able to easily pull the code and build a local version yourself then run against your Kafka topic, or is that too tricky in your context?

**Comment by jskeet** [comment]
I've merged the PR now, but if you're able to test it without a release, that would be great. Don't worry if not though.

**Comment by jskeet** [comment]
Excellent! Thanks very much. In that case, I'll try to do a release in the next day or two.

**Comment by jskeet** [comment]
This has now been released in 2.3.0. Thanks for raising it.

**Comment by arekpalinski** [comment]
Thank you! I've just updated the packages and can confirm that the problem is solved.
